internal/pkg/action/operator_uninstall.go

```diff
@@ -86,45 +87,39 @@ func (u *OperatorUninstall) Run(ctx context.Context) error {
 		return err
 	}
 
-	if csv != nil {
-		// Ensure CustomResourceDefinitions are deleted next, so that the operator
-		// has a chance to handle CRs that have finalizers.
-		if u.DeleteCRDs {
-			crds := getCRDs(csv)
-			if err := u.deleteObjects(ctx, crds...); err != nil {
-				return err
-			}
-		}
-
-		// OLM puts an ownerref on every namespaced resource to the CSV,
-		// and an owner label on every cluster scoped resource. When CSV is deleted
-		// kube and olm gc will remove all the referenced resources.
-		if err := u.deleteObjects(ctx, csv); err != nil {
-			return err
-		}
+	operands, err := u.operandLister.Run(ctx, u.Package)
+	if err != nil {
+		return fmt.Errorf("listing operands associated with operator %s: %s", u.Package, err)
 	}
+	u.Logf("found %d operands associated with operator %s", len(operands.Items), u.Package)
+	u.Logf("operand deletion strategy: %s", u.OperandStrategy.String())
 
-	if u.DeleteOperatorGroups {
-		subs := v1alpha1.SubscriptionList{}
-		if err := u.config.Client.List(ctx, &subs, client.InNamespace(u.config.Namespace)); err != nil {
-			return fmt.Errorf("list subscriptions: %v", err)
+	switch u.OperandStrategy {
+	case operand.Cancel:
+		if len(operands.Items) > 0 {
+			return fmt.Errorf("deleting an operator with operands on-cluster requires a operand deletion strategy to be set")
 		}
-		// If there are no subscriptions left, delete the operator group(s).
-		if len(subs.Items) == 0 {
-			ogs := v1.OperatorGroupList{}
-			if err := u.config.Client.List(ctx, &ogs, client.InNamespace(u.config.Namespace)); err != nil {
-				return fmt.Errorf("list operatorgroups: %v", err)
-			}
-			for _, og := range ogs.Items {
-				og := og
-				if len(u.DeleteOperatorGroupNames) == 0 || contains(u.DeleteOperatorGroupNames, og.GetName()) {
-					if err := u.deleteObjects(ctx, &og); err != nil {
-						return err
-					}
-				}
```

Comment:
I'm not convinced we should get rid of this. It seems pretty useful, and I think is somewhat orthogonal to the rest of this PR. Could we keep this for this PR and have a more focused issue/PR for operator group deletion in a follow-up?

Reply:
We could break that out into a separate PR, sure. 

I feel pretty strongly that we should only delete the minimal subset of items for this command for the initial downstream build, mainly because operatorgroups are not well understood by most OLM users. If a user does understand operatorgroups, they should feel comfortable deleting them manually, even though it can have implications for subsequent installations.

The scenario I'd like to avoid is having a user inadvertently delete the global operatorgroup in the `openshift-operators` namespace with the `uninstall` command and then have all subsequent operator installations fail.

Reply:
> mainly because operatorgroups are not well understood by most OLM users

This cuts both ways. With `install` creating an operatorgroup (if necessary), and `uninstall` deleting the operatorgroup (when possible), the commands have symmetry and the existing UX means users _can_ be completely unaware of what they are and how they work and still interact with OLM.

> The scenario I'd like to avoid is having a user inadvertently delete the global operatorgroup in the openshift-operators namespace with the uninstall command and then have all subsequent operator installations fail.

This is not a scenario I had considered, but I agree it should be. Is it currently possible for a user with the necessary RBAC permissions to delete the global operatorgroup in the `openshift-operators` namespace permanently? Or will some controller re-create it when it finds that it is missing?